### PR TITLE
Unit Tests

### PR DESCRIPTION
**Pull request by iruzevic**
- CiExclude
- Helpers
- Login
- ModifyAdminAppearance
- I18n
- GitIgnore
- Readme
- ServicesExample

### CODE REVIEW

File: tests/Helpers/ComponentHelpersTest.php [hunk 1]
```diff
@@ -31,6 +30,13 @@ function ($argument) {
 	Components::ensureString();
 })->throws(\ArgumentCountError::class);
 
+error_log( print_r( ( Components::ensureString(0) ), true ) );
+
+test('Throws exception if any other argument is passed other than string or array', function () {
```
Comment (iruzevic) [inline tests/Helpers/ComponentHelpersTest.php:35]: @dingo-d Can you please help me why is this failing?

"Call to undefined function esc_html__()"

File: tests/Helpers/ComponentHelpersTest.php [hunk 2]
```diff
@@ -31,6 +30,13 @@ function ($argument) {
 	Components::ensureString();
 })->throws(\ArgumentCountError::class);
 
+error_log( print_r( ( Components::ensureString(0) ), true ) );
```
Comment (dingo-d) [inline tests/Helpers/ComponentHelpersTest.php:33]: Rogue error log. If you want to see what it throws during testing you'll need to put the error log inside the test (or just `print_r` or `var_dump`). I mean the best way is to use the debugger to see what is happening during the test.

File: tests/Helpers/ComponentHelpersTest.php (hunk 1 above)
Comment (dingo-d) [reply tests/Helpers/ComponentHelpersTest.php:35]: It shouldn't, seeing how `setupMocks` is called before all tests. Maybe it should be `beforeEach(function() { setupMocks(); });` 🤷🏼‍♂️ 

But anyhow, this case should be covered with the test on line 22. If you look at it closely it will take an arbitrary number of arguments, and test that the type error is thrown, using the `errorStringArguments` dataset, defined in `Arguments.php`.

Comment (dingo-d) [reply tests/Helpers/ComponentHelpersTest.php:35]: More to the point I just ran the tests and it's the test on line 22 that is faulty. It should be:

```php
test('Throws type exception if wrong argument type is passed to ensureString', function ($argument) {
		Components::ensureString($argument);
	})
	->throws(ComponentException::class)
	->with('errorStringArguments');
```

And I found an edge case. When you have a standard object created using `new class{}`, you get

```bash
• Tests\Helpers\ComponentHelpersTest > Throws type exception if wrong argument type is passed to ensureString with (class@anonymous\/Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/tests/Datasets/Arguments.php:9$1dc Object ())
  Failed asserting that exception of type "Error" matches expected exception "EightshiftLibs\Exception\ComponentException". Message was: "Object of class class@anonymous could not be converted to string" at
/Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/src/Exception/ComponentException.php:33
  /Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/src/Helpers/Components.php:48
  /Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/tests/Helpers/ComponentHelpersTest.php:23
  /Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/vendor/pestphp/pest/src/Factories/TestCaseFactory.php:142
  /Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/vendor/pestphp/pest/src/Concerns/TestCase.php:175
  /Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/vendor/pestphp/pest/src/Support/ExceptionTrace.php:28
  /Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/vendor/pestphp/pest/src/Concerns/TestCase.php:176
  /Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/vendor/pestphp/pest/src/Concerns/TestCase.php:164
  /Users/denis.zoljom/Projects/Infinum/Eightshift/eightshift-libs/vendor/pestphp/pest/src/Console/Command.php:130
```

That's because this on-the-fly object in the dataset doesn't have `__toString` magic method. So we found an edge case that should be covered in the code \o/ 🎉  😄

**Review by IvanGrginovInf — APPROVED** [review]

File: tests/Helpers/ComponentHelpersTest.php (hunk 1 above)
Comment (iruzevic) [reply tests/Helpers/ComponentHelpersTest.php]: fixed

**Review by kmaric69 — APPROVED** [review]

**Review by IvanGrginovInf — APPROVED** [review]

File: src/Exception/ComponentException.php [hunk 3]
```diff
@@ -19,20 +19,24 @@ final class ComponentException extends \InvalidArgumentException implements Gene
...
+			$output = sprintf(
+				/* translators: %1$s is replaced with the name of the variable, and %2$s with its type. */
 				\esc_html__('%1$s variable is not a string or array but rather %2$s', 'eightshift-libs'),
 				$variable,
 				gettype($variable)
-			)
-		);
+			);
+		} else {
+			$output = \esc_html__('Object can\'t be converted to string. Please provide only string or array.', 'eightshift-libs');
```
Comment (dingo-d) [inline src/Exception/ComponentException.php:36]:
```suggestion
			$output = \esc_html__('Object couldn\'t be converted to string. Please provide only string or array.', 'eightshift-libs');
```

We should see what the output of the test is for the `ensureString` if we pass the object with `__toString` magic method in it 😄

Comment (dingo-d) [reply src/Exception/ComponentException.php:36]: Probably nothing tho, since the `ensureString` will work only with strings and arrays 🤔

File: src/Helpers/Components.php [hunk 4]
```diff
@@ -129,11 +128,23 @@ public static function render(string $component, array $attributes = [], string
...
+			$defaultAttributes = [];
+			foreach ($manifest['attributes'] as $itemKey => $itemValue) {
+				if (isset($itemValue['default'])) {
+					$defaultAttributes[$itemKey] = $itemValue['default'];
+				}
+			}
+
+			$attributes = array_merge($defaultAttributes, $attributes);
+		}
+
```
Comment (dingo-d) [inline src/Helpers/Components.php:141]: This doesn't seem to be covered with tests. Can you write a test for this as well?

File: tests/CiExclude/CiExcludeCliTest.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+test('CiExclude CLI command will correctly copy the ci-exclude text file with defaults', function () {
+	$ciexclude = $this->ciexclude;
+	$ciexclude([], $ciexclude->getDevelopArgs([]));
+	
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedExclude = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/ci-exclude.txt');
+
+	$this->assertStringContainsString('eightshift-boilerplate', $generatedExclude);
```
Comment (dingo-d) [inline tests/CiExclude/CiExcludeCliTest.php:43]: Try to add an 'unhappy path' value here, like

```php
$this->assertStringNotContainsString('footer.php', $generatedExclude);
```

File: tests/CiExclude/CiExcludeCliTest.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+test('CiExclude CLI command will correctly copy the ci-exclude file with set arguments', function () {
+	$ciexclude = $this->ciexclude;
+	$ciexclude([], [
+		'root' => './',
+		'project_name' => 'coolPlugin',
+		'project_type' => 'plugin',
+	]);
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedExclude = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/ci-exclude.txt');
```
Comment (dingo-d) [inline tests/CiExclude/CiExcludeCliTest.php:74]: While this on its own is a good indicator that the file exists, it's a bit fragile. Sure, if it fails, you can say: something's wrong. And that's a good indicator, but we have a method that will test if the file has been created:

```php
test('CiExclude CLI command will correctly copy the ci-exclude file in the custom folder with set arguments', function () {
	$ciexclude = $this->ciexclude;
	$ciexclude([], [
		'root' => './test',
	]);

	// Check the output dir if the generated method is correctly generated.
	$this->assertFileExists(dirname(__FILE__, 3) . '/cliOutput/test/ci-exclude.txt');
});
```

🎉

File: tests/GitIgnore/GitIgnoreCliTest.php [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+test('GitIgnore CLI command will correctly copy the .gitignore file with defaults', function () {
+	$gitignore = $this->gitignore;
+	$gitignore([], $gitignore->getDevelopArgs([]));
+	
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedIgnore = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/.gitignore');
+
+	$this->assertStringContainsString('wp-admin', $generatedIgnore);
```
Comment (dingo-d) [inline tests/GitIgnore/GitIgnoreCliTest.php:43]: One unhappy path assertion, please 😄

File: tests/GitIgnore/GitIgnoreCliTest.php [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+});
+
+test('GitIgnore CLI command will correctly copy the .gitignore file in the custom folder with set arguments', function () {
+	$gitignore = $this->gitignore;
+	$gitignore([], [
+		'root' => './test',
+	]);
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedIgnore = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/test/.gitignore');
```
Comment (dingo-d) [inline tests/GitIgnore/GitIgnoreCliTest.php:60]: Use `assertFileExists` assertion here as well 🙂

File: tests/I18n/I18nCliTest.php [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+test('I18n CLI command will correctly copy the I18n class with defaults', function () {
+	$i18n = $this->i18n;
+	$i18n([], []);
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedMain = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/src/I18n/I18n.php');
+
+	$this->assertStringContainsString('class I18n implements ServiceInterface', $generatedMain);
+	$this->assertStringContainsString('@package EightshiftBoilerplate\I18n', $generatedMain);
+	$this->assertStringContainsString('namespace EightshiftLibs\I18n', $generatedMain);
```
Comment (dingo-d) [inline tests/I18n/I18nCliTest.php:44]: Assertion that something is not in the file 🙏🏼

File: tests/I18n/I18nExampleTest.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+test('Registering load_theme_textdomain works', function () {
+
+	// Set up a side-affect.
+	putenv('I18N_ABSTRACTED=false');
+
+
+	\Mockery::mock('alias:EightshiftBoilerplate\Config\Config')
+	->shouldReceive('getProjectName', 'getProjectPath')
+	->andReturn('CoolProject', 'projectPath');
```
Comment (dingo-d) [inline tests/I18n/I18nExampleTest.php:36]: Can you just indent this by 1 tab, so it's a bit more readable? Thanks!

File: tests/Login/LoginCliTest.php [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+test('Login CLI command will correctly copy the Login class with defaults', function () {
+	$login = $this->login;
+	$login([], []);
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedLogin = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/src/Login/Login.php');
+
+	$this->assertStringContainsString('class Login implements ServiceInterface', $generatedLogin);
+	$this->assertStringContainsString('@package EightshiftBoilerplate\Login', $generatedLogin);
+	$this->assertStringContainsString('namespace EightshiftLibs\Login', $generatedLogin);
```
Comment (dingo-d) [inline tests/Login/LoginCliTest.php:44]: Unhappy path

File: tests/Main/MainCliTest.php [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+test('Main CLI command will correctly copy the Main class with defaults', function () {
+	$main = $this->main;
+	$main([], []);
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedMain = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/src/Main/Main.php');
+
+	$this->assertStringContainsString('class Main extends AbstractMain', $generatedMain);
+	$this->assertStringContainsString('@package EightshiftBoilerplate\Main', $generatedMain);
+	$this->assertStringContainsString('namespace EightshiftLibs\Main', $generatedMain);
```
Comment (dingo-d) [inline tests/Main/MainCliTest.php:44]: Unhappy path

File: tests/ModifyAdminAppearance/ModifyAdminAppearanceCliTest.php [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+test('ModifyAdminAppearance CLI command will correctly copy the ModifyAdminAppearance class with defaults', function () {
+	$modifyAdminAppearance = $this->modifyAdminAppearance;
+	$modifyAdminAppearance([], []);
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedModifyAdminAppearance = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/src/ModifyAdminAppearance/ModifyAdminAppearance.php');
+
+	$this->assertStringContainsString('class ModifyAdminAppearance implements ServiceInterface', $generatedModifyAdminAppearance);
+	$this->assertStringContainsString('@package EightshiftBoilerplate\ModifyAdminAppearance', $generatedModifyAdminAppearance);
+	$this->assertStringContainsString('namespace EightshiftLibs\ModifyAdminAppearance', $generatedModifyAdminAppearance);
```
Comment (dingo-d) [inline tests/ModifyAdminAppearance/ModifyAdminAppearanceCliTest.php:44]: Unhappy path

File: tests/Readme/ReadmeCliTest.php [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+test('Readme CLI command will correctly copy the readme in the custom folder with set arguments', function () {
+	$readme = $this->readme;
+	$readme([], [
+		'root' => './test',
+	]);
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedExclude = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/test/README.md');
+
+	$this->assertStringContainsString('This is the official repository of the {Project name}.', $generatedExclude);
```
Comment (dingo-d) [inline tests/Readme/ReadmeCliTest.php:62]: File exists assertion can be used here

File: tests/Services/ServicesCliTest.php [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+test('Services CLI command will correctly copy the Services class with defaults', function () {
+	$services = $this->services;
+	$services([], $services->getDevelopArgs([]));
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedMain = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/src/TestFolder/TMP/TestTest.php');
+
+	$this->assertStringContainsString('class TestTest implements ServiceInterface', $generatedMain);
+	$this->assertStringContainsString('namespace EightshiftLibs\TestFolder\TMP', $generatedMain);
+	$this->assertStringContainsString('@package EightshiftBoilerplate\TestFolder\TMP', $generatedMain);
```
Comment (dingo-d) [inline tests/Services/ServicesCliTest.php:44]: Unhappy path

File: tests/Services/ServicesCliTest.php [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+	$services = $this->services;
+	$services([], [
+		'namespace' => 'CoolTheme',
+		'folder' => 'FolderName',
+		'file_name' => 'FileName',
+	]);
+
+	// Check the output dir if the generated method is correctly generated.
+	$generatedMain = file_get_contents(dirname(__FILE__, 3) . '/cliOutput/src/FolderName/FileName.php');
+
```
Comment (dingo-d) [inline tests/Services/ServicesCliTest.php:57]: File exists assertion can be used here.

**Review by dingo-d — CHANGES_REQUESTED** [review]
Left some minor comments. There is one weird output in the tests, not sure where this is coming from:

```bash
#!/usr/bin/env sh

set -e

function build() {
	cd 'wp-content/themes/eightshift-boilerplate';
	npm install
	composer install --no-dev
	npm run build
}

build
```

Great job 👍🏼

File: src/Helpers/Components.php (hunk 4 above)
Comment (iruzevic) [reply src/Helpers/Components.php:141]: Yes but in the next test batch. I will open an new pr for this

File: tests/CiExclude/CiExcludeCliTest.php (hunk 6 above)
Comment (iruzevic) [reply tests/CiExclude/CiExcludeCliTest.php:74]: awesome tnx

**Comment by iruzevic** [comment]
Fixed all

File: tests/ModifyAdminAppearance/ModifyAdminAppearanceExampleTest.php [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+	$this->assertStringContainsString('blue', $output);
+});
+
+test('Asserts if wp_get_environment_type is empty', function () {
+
+	Functions\when('wp_get_environment_type')->justReturn(null);
+
+	$output = $this->modifyAdminAppearance->adminColor();
+
+	$this->assertStringContainsString('fresh', $output);
```
Comment (dingo-d) [inline tests/ModifyAdminAppearance/ModifyAdminAppearanceExampleTest.php:43]: I'd add something like `$this->assertStringNotContainsString('sunrise', $output);` either here or in the above test. Maybe even in the above test to confirm the color scheme is correct in case of a specific env.

File: tests/Services/ServiceExampleTest.php [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+});
+
+afterEach(function() {
+	Monkey\tearDown();
+});
+
+test('Service Example test register method', function () {
+	$this->service->register();
+
+	$this->assertTrue(true);
```
Comment (dingo-d) [inline tests/Services/ServiceExampleTest.php:25]: This is not a valid test 👀

File: src/Helpers/Components.php (hunk 4 above)
Comment (dingo-d) [reply src/Helpers/Components.php:141]: Usually, when you add a feature you need to cover it with tests in the same branch 🙈

**Review by dingo-d — APPROVED** [review]

**Comment by iruzevic** [comment]
💣
